core/src/wsv.rs

```diff
@@ -393,30 +395,39 @@ impl WorldStateView {
         self.asset(id).map_err(Into::into)
     }
 
-    // TODO: There could be just this one method `blocks` instead of
-    // `blocks_from_height` and `blocks_after_height`. Also, this
-    // method would return references instead of cloning blockchain
-    // but comes with the risk of deadlock if consumer of the iterator
-    // stores references to blocks
-    /// Returns iterator over blockchain blocks
-    ///
-    /// **Locking behaviour**: Holding references to blocks stored in the blockchain can induce
-    /// deadlock. This limitation is imposed by the fact that blockchain is backed by [`dashmap::DashMap`]
-    #[inline]
-    pub fn blocks(&self) -> crate::block::ChainIterator {
-        self.blocks.iter()
+    /// Load all blocks in the block chain from disc and clone them for use.
+    #[allow(clippy::expect_used)]
+    pub fn all_blocks_by_value(&self) -> Vec<VersionedCommittedBlock> {
+        let block_count = self.block_hashes.borrow().len() as u64;
+        (1..=block_count)
+            .map(|height| {
+                self.kura
+                    .get_block_by_height(height)
+                    .expect("Failed to load block.")
+            })
+            .map(|block| VersionedCommittedBlock::clone(&block))
+            .collect()
     }
 
     /// Return a vector of blockchain blocks after the block with the given `hash`
-    pub fn blocks_after_hash(
+    pub fn block_hashes_after_hash(
         &self,
         hash: HashOf<VersionedCommittedBlock>,
-    ) -> Vec<VersionedCommittedBlock> {
-        self.blocks
-            .iter()
-            .skip_while(move |block_entry| block_entry.value().header().previous_block_hash != hash)
-            .map(|block_entry| block_entry.value().clone())
-            .collect()
+    ) -> Vec<HashOf<VersionedCommittedBlock>> {
```

Comment:
I don't think so since the data is guarded by mutex. In general I think it's good lock, produce copies of data, unlock, return the copied data. It is a style less likely to cause bugs.